### DISCUSSION TITLE
日志写为json格式的问题

### DISCUSSION THREAD

**Discussion by lf2186** [bfenetworks/bfe#949] [Q&A (问题和回答)] [answered]
hi,

     你好，

想把日志记录改为json格式，如下，修改mod_access配置文件：
RequestTemplate = "\{ \"ReqLog_type\": \"REQUEST_LOG\", \"local_time\": \"$time\", \"clientip\": \"$remote_client_addr\", "\http_me
thod\": \"$req_method\", \"http_scheme\": \"$req_schema\", \"host\": \"$host\", \"url\": \"$req_url\", \"status\": \"$status_code\",
 \"reqbodylen\": \"$req_body_len\", \"respbodylen\": \"$res_body_len\", \"time_all\": \"$all_time\", \"time_connect_rs\": \"$connect
_time\", \"product\": \"$product\", \"backend\": \"$backend\", \"pro_version\": \"$res_proto\", \"req_seq\": \"$req_nth\", \"clientp
ort\": \"$remote_client_port\", \"user_agent\": \"${User-Agent}req_header\", \"error\": \"$error\", \"xff\": \"${X-Forwarded-For}req
_header\", \"xrealip\": \"${X-Real-IP}req_header\", \"rrdt\": \"$read_req_duration\", \"pdt\": \"$proxy_delay\", \"wst\": \"$write_s
erve_time\", \"rdt\": \"$response_duration\", \"cdt\": \"$cluster_duration\", \"lbdt\": \"$last_backend_duration\", \"rst\": \"$read
write_serve_time\", \"ssst\": \"$since_ses_start_time\" \}"
<img width="1183" alt="image" src="https://user-images.githubusercontent.com/4362152/148937755-b252824b-0dad-467f-a0d2-c8088c880694.png">

但是启动bfe报错：
[01/11/22 19:27:33] [EROR] Err in module.init() for mod_access [mod_access: cond load err conf/mod_access/mod_access.conf:9:21: unknown escape sequence (and 1 more errors)]
[01/11/22 19:27:33] [EROR] StartUp(): bfeServer.InitModules():mod_access: cond load err conf/mod_access/mod_access.conf:9:21: unknown escape sequence (and 1 more errors)
[01/11/22 19:27:33] [EROR] main(): bfe_server.StartUp(): mod_access: cond load err conf/mod_access/mod_access.conf:9:21: unknown escape sequence (and 1 more errors)

怎么解决？？

**Comment by liuqing6767** [comment, marked as answer]
基于问题描述，做了一个配置：
```
[Template]
RequestTemplate = "{\"ReqLog_type\":\"REQUEST_LOG\",\"local_time\":\"$time\",\"clientip\":\"$ses_clientip\" ,\"http_method\":\"?req_method\",\"http_scheme\":\"?req_schema\",\"host\":\"$host\",\"url\":\"$url\",\"status\":\"$status_code\",\"reqbodylen\":\"$req_body_len\",\"respbodylen\":\"$res_body_len\",\"time_all\":\"$all_time\",\"time_connect_rs\":\"$connect_time\",\"product\":\"$product\",\"backend\":\"$backend\",\"pro_version\":\"$res_proto\",\"req_seq\":\"$req_nth\",\"clientport\":\"?remote_client_port\",\"user_agent\":\"User-Agent $req_header\",\"error\":\"$error\",\"xff\":\"X-Forwarded-For req_header\",\"xrealip\":\"${X-Real-IP}req_header\",\"rrdt\":\"$read_req_duration\",\"pdt\":\"$proxy_delay\",\"wst\":\"$write_serve_time\",\"rdt\":\"$response_duration\",\"cdt\":\"$cluster_duration\",\"lbdt\":\"$last_backend_duration\",\"rst\":\"$readwrite_serve_time\",\"ssst\":\"$since_ses_start_time\"}"
```
可得到如下日志：
```
{"ReqLog_type":"REQUEST_LOG","local_time":"2022-01-12 16:09:27","clientip":"127.0.0.1" ,"http_method":"?req_method","http_scheme":"?req_schema","host":"127.0.0.1:8080","url":"/aaa","status":"500","reqbodylen":"0","respbodylen":"0","time_all":"1","time_connect_rs":"","product":"-","backend":",,,","pro_version":"","req_seq":"1","clientport":"?remote_client_port","user_agent":"curl/7.77.0","error":"BK_FIND_PRODUCT,no product found","xff":"xxxxdfafafaf","xrealip":"-","rrdt":"0","pdt":"-","wst":"0","rdt":"0","cdt":"0","lbdt":"0","rst":"-","ssst":"1"}
```

问题中的示例：
- 不是一个合法的json字符串
- 加入了自定义的变量（这个不是问题）

配置项具体构造过程建议如下步骤：
- 构造json格式的配置模板，形如： 
```
{
	"ReqLog_type": "REQUEST_LOG",
	"local_time": "$time"
}
```
- 压缩json
- 转义json 得到 t
- 配置项为：  `RequestTemplate = "{t}"`

（json操作可以用 https://www.bejson.com/）

另外，请关注：如果变量计算结果存在 双引号，因为 日志生产时不会转义，可能出现非法json字符串。

请按建议步骤再进行尝试或反馈。
